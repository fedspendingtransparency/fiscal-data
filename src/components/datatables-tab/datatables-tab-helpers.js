export const fileSizeTranslator = fileSize => {
  const fileSizeNumber = Number(fileSize);
  if (fileSizeNumber < 1000000) {
    const theNumber = Math.round(fileSizeNumber / 1000);
    if (theNumber === 1000) {
      return '1 MB';
    } else {
      return theNumber.toString() + ' KB';
    }
  } else {
    return Math.round(fileSizeNumber / 1000000).toString() + ' MB';
  }
};

/*
 *             ~~~~~~~fileSizeTranslator2~~~~~~~
 *                   This time it's binary
 */
export const fileSizeTranslator2 = fileSize => {
  if (fileSize && !isNaN(fileSize)) {
    const fileSizeNumber = Number(fileSize);
    const fileSizes = ['B', 'KB', 'MB', 'GB'];
    const kb = 1024;
    let curMultiplier = 1;
    for (let i = 0, il = fileSizes.length; i < il; i++) {
      const curThreshold = curMultiplier * kb;
      if (curThreshold > fileSizeNumber) {
        return Math.round(fileSizeNumber / curMultiplier) + ` ${fileSizes[i]}`;
      }
      curMultiplier = curThreshold;
    }
  }

  return null;
};

export const makeTheDataArray = apisArray => {
  const theDataArray = [];
  apisArray.forEach(api => {
    // todo - DTG-2209 will provide the file sizes.
    const dataObj = {
      name: api.tableName,
      description: api.tableDescription,
      rowCount: api.rowCount,
      rowDescription: api.rowDefinition,
    };
    theDataArray.push(dataObj);
  });
  return theDataArray;
};
