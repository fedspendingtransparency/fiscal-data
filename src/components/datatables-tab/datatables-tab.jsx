import React from 'react';
import DtgTable from '../dtg-table/dtg-table';
import { makeTheDataArray } from './datatables-tab-helpers';

export default function DataTablesTab({ apis, datasetName }) {
  const theData = makeTheDataArray(apis);
  const columnConfig = [
    {
      property: 'name',
      name: 'Table Name',
      order: 1,
      width: 12,
    },
    {
      property: 'description',
      name: 'Description',
      order: 2,
      width: 38,
    },
    {
      property: 'rowDescription',
      name: 'Row Description',
      order: 3,
      width: 38,
    },
    {
      property: 'rowCount',
      name: 'Row Count',
      order: 4,
      width: 12,
    },
  ];

  const tableProps = {
    data: theData,
    columnConfig,
    width: 1104,
    tableName: 'Datatables Tab',
    aria: { 'aria-label': `${datasetName} data tables` },
  };

  return <DtgTable tableProps={tableProps} />;
}
