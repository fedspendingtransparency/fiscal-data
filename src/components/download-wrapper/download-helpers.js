import { postAPI } from '../../utils/api-utils';
import { format } from 'date-fns';
import { fileSizeTranslator2 } from '../datatables-tab/datatables-tab-helpers';

export const getDateRangeForFiltration = (dateRange, endpoint) => {
  // redemption_tables is a one-off scenario where the date string needs to be YYYY-MM.
  if (endpoint && endpoint.indexOf('redemption_tables') > -1) {
    return {
      from: format(dateRange.from, 'yyyy-MM'),
      to: format(dateRange.to, 'yyyy-MM'),
    };
  }

  return {
    from: format(dateRange.from, 'yyyy-MM-dd'),
    to: format(dateRange.to, 'yyyy-MM-dd'),
  };
};

const filenameDate = filtrationDate => {
  return filtrationDate.replace(/-/g, '');
};

export const getFilenameForSaveAs = (dateFilter, api, format) => {
  const dateAppendix = dateFilter ? '_' + filenameDate(dateFilter.from) + '_' + filenameDate(dateFilter.to) : '';
  return api.downloadName + dateAppendix + '.' + format;
}

export const replaceNbsps = str => {
  const re = new RegExp(String.fromCharCode(160), 'g'); // global search for " "
  return str.replace(re, ' ');
};

const makeHeadRequests = async (filePath, fetchHeadOption, failedRequestsNbr) => {
  if (failedRequestsNbr <= 3) {
    return postAPI(filePath, fetchHeadOption)
      .then(res => {
        let returnVal = null;
        if (res && res.headers && res.headers.get) {
          const contentType = res.headers.get('Content-Type');
          const fileSize = res.headers.get('Content-Length');
          if (fileSize && contentType === 'application/zip') {
            returnVal = fileSizeTranslator2(fileSize);
          }
          // Uncomment the below for testing file sizes in local
          // else {
          //   returnVal = Math.round(Math.random() * 100) + ' MB';
          // }
        }
        return returnVal;
      })
      .catch(async e => {
        await makeHeadRequests(filePath, fetchHeadOption, ++failedRequestsNbr);
      });
  }
  return null;
};

export const constructDownloadFileName = (dateRange, selectedTable) => {
  if (dateRange?.from && dateRange?.to && selectedTable?.downloadName) {
    console.log('selectedTable?.downloadName:: ', selectedTable?.downloadName);
    const from = format(dateRange.from, 'yyyyMMdd');
    const to = format(dateRange.to, 'yyyyMMdd');
    return selectedTable.downloadName + '_' + from + '_' + to;
  }
};
