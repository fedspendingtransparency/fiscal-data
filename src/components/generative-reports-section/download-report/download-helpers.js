import { postAPI } from '../../../utils/api-utils';
import { fileSizeTranslator2 } from '../../datatables-tab/datatables-tab-helpers';

export const getFileSize = async (filePath, failedRequestsNbr = 0) => {
  const fetchHeadOption = {
    method: 'HEAD',
  };

  if (failedRequestsNbr < 3) {
    return postAPI(filePath, fetchHeadOption)
      .then(res => {
        let returnVal = null;
        if (res && res.headers && res.headers.get) {
          const fileSize = res.headers.get('Content-Length');
          if (fileSize) {
            returnVal = fileSizeTranslator2(fileSize);
          }
          // Uncomment the below for testing file sizes in local
          // else {
          //   returnVal = Math.round(Math.random() * 100) + ' MB';
          // }
        }
        return returnVal;
      })
      .catch(async e => {
        await getFileSize(filePath, ++failedRequestsNbr);
      });
  } else {
    return null;
  }
};
