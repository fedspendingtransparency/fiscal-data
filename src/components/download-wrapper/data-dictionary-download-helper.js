import { fileSizeTranslator } from '../datatables-tab/datatables-tab-helpers';
import fileDownload from 'js-file-download';
import { stringify } from 'csv-stringify/sync';
import { replaceNbsps } from './download-helpers';

export const convertDataDictionaryToCsv = dataset => {
  console.log('dataset: ', dataset);
  const apis = dataset.apis;
  if (apis && apis[0] && apis[0].fields && apis[0].fields.length) {
    const allTableFields = apis.reduce((accumulated, current) => {
      return accumulated.concat(addDataset(current.fields, dataset));
    }, []);
    return replaceNbsps(
      stringify(allTableFields, {
        header: true,
        columns: [
          {
            key: 'dataset',
            header: 'dataset',
          },
          {
            key: 'tableName',
            header: 'data_table_name',
          },
          {
            key: 'columnName',
            header: 'field_name',
          },
          {
            key: 'prettyName',
            header: 'display_name',
          },
          {
            key: 'definition',
            header: 'description',
          },
          {
            key: 'dataType',
            header: 'data_type',
          },
          {
            key: 'isRequired',
            header: 'is_required',
          },
        ],
      })
    );
  } else {
    return '';
  }
};

export const replaceNbsps = str => {
  const re = new RegExp(String.fromCharCode(160), 'g'); // global search for " "
  return str.replace(re, ' ');
};

const addDataset = (fields, dataset) => {
  return fields.map(field => Object.assign({ dataset: dataset.name }, field));
};

export const suggestDictionaryDownloadName = datasetName => {
  return datasetName.replace(/[^a-zA-z ]/g, '') + ' Data Dictionary.csv';
};

export const calcDictionaryDownloadSize = csvData => {
  return fileSizeTranslator(1000 + csvData.length);
};

export const triggerDataDictionaryDownload = (csvData, datasetName) => {
  return fileDownload(csvData, suggestDictionaryDownloadName(datasetName));
};
