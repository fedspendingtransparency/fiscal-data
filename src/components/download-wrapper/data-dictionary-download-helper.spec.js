import { calcDictionaryDownloadSize, convertDataDictionaryToCsv, prettySize, suggestDictionaryDownloadName } from './data-dictionary-download-helper';
import { replaceNbsps } from './download-helpers';
import { fileSizeTranslator2 } from '../datatables-tab/datatables-tab-helpers';
jest.mock('../datatables-tab/datatables-tab-helpers', () => {
  const actual = jest.requireActual('../datatables-tab/datatables-tab-helpers');
  return {
    ...actual,
    fileSizeTranslator2: jest.fn(),
  };
});

describe('Data Dictionary Helper function collection', () => {
  it('removes unwanted non-breaking spaces', () => {
    const words = ['This', 'is', 'a', 'sentence.'];
    const sentenceWithSpaces = 'This is a sentence.';
    const noBreakSentence = words.join(String.fromCharCode(160));
    expect(noBreakSentence).not.toEqual(sentenceWithSpaces);
    expect(replaceNbsps(noBreakSentence)).toEqual(sentenceWithSpaces);
  });

  it('correctly applies the csv formatting to the data dictionary', () => {
    const dataset = {
      datasetId: '015-TLP-2014Q3-065',
      name: 'Debt to the Nickel',
      apis: [
        {
          tableName: 'table1',
          fields: [
            {
              tableName: 'table1',
              columnName: 'reporting_date',
              prettyName: 'Calendar Date',
              definition: 'Reporting date for the data',
              dataType: 'DATE',
              isRequired: 'yes',
            },
            {
              tableName: 'table1',
              columnName: 'reporting_date',
              prettyName: 'Calendar Date',
              definition: 'Reporting date for the data',
              dataType: 'DATE',
              isRequired: 'yes',
            },
          ],
        },
        {
          tableName: 'table2',
          fields: [
            {
              tableName: 'table2',
              columnName: 'reporting_date',
              prettyName: 'Calendar Date',
              definition: 'Reporting date for the data',
              dataType: 'DATE',
              isRequired: 'yes',
            },
            {
              tableName: 'table2',
              columnName: 'reporting_date',
              prettyName: 'Calendar Date',
              definition: 'Reporting date for the data',
              dataType: 'DATE',
              isRequired: 'yes',
            },
          ],
        },
      ],
    };
    expect(convertDataDictionaryToCsv(dataset)).toMatchSnapshot();
  });

  it('suggests a correctly formatted Data Dictionary download filename', () => {
    expect(suggestDictionaryDownloadName('Debt to the Nickel')).toEqual('Debt to the Nickel Data Dictionary.csv');
  });

  it('calculates the approx filesize size of the CSV download', () => {
    expect(calcDictionaryDownloadSize('0123456789')).toEqual('1 KB');
  });
  describe('prettySize', () => {
    beforeEach(() => {
      fileSizeTranslator2.mockReset();
    });

    it('uses fileSizeTranslator2 on string data', () => {
      const str = 'hello';
      const expectedBytes = new Blob([str]).size;
      fileSizeTranslator2.mockReturnValue('5 B');

      const result = prettySize(str);

      expect(fileSizeTranslator2).toHaveBeenCalledWith(expectedBytes);
      expect(result).toBe('5 B');
    });

    it('uses fileSizeTranslator2 on Blob data', () => {
      const blob = new Blob(['abcdef']);
      fileSizeTranslator2.mockReturnValue('6 B');

      const result = prettySize(blob);

      expect(fileSizeTranslator2).toHaveBeenCalledWith(blob.size);
      expect(result).toBe('6 B');
    });

    it('uses fileSizeTranslator2 on array of rows', () => {
      const rows = [
        ['a', 'b'],
        ['c', 'd'],
      ];
      const csvString = 'a,b\nc,d';
      fileSizeTranslator2.mockReturnValue('7 B');
      const result = prettySize(rows);
      expect(fileSizeTranslator2).toHaveBeenCalledWith(new Blob([csvString]).size);
      expect(result).toBe('7 B');
    });

    it('falls back to "-" for unsupported types', () => {
      expect(prettySize(123)).toBe('-');
    });

    it('falls back to "-" if translator returns falsy', () => {
      fileSizeTranslator2.mockReturnValue('');
      expect(prettySize('x')).toBe('-');
    });
  });
});
