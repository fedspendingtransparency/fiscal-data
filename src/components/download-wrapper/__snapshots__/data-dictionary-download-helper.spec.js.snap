// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Dictionary Helper function collection correctly applies the csv formatting to the data dictionary 1`] = `
"dataset,data_table_name,field_name,display_name,description,data_type,is_required
Debt to the Nickel,table1,reporting_date,Calendar Date,Reporting date for the data,DATE,yes
Debt to the Nickel,table1,reporting_date,Calendar Date,Reporting date for the data,DATE,yes
Debt to the Nickel,table2,reporting_date,Calendar Date,Reporting date for the data,DATE,yes
Debt to the Nickel,table2,reporting_date,Calendar Date,Reporting date for the data,DATE,yes
"
`;
